boot/seal.go

```diff
@@ -95,40 +98,101 @@ func sealKeyToModeenv(key secboot.EncryptionKey, model *asserts.Model, modeenv *
-		TPMLockoutAuthFile:   filepath.Join(InstallHostFDEDataDir, "tpm-lockout-auth"),
+
+	// the boot chains we seal the fallback object to
+	rpbc := toPredictableBootChains(recoveryBootChains)
+
+	authKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return fmt.Errorf("cannot generate key for signing dynamic authorization policies: %v", err)
 	}
-	// finally, seal the key
-	if err := secbootSealKey(key, sealKeyParams); err != nil {
-		return fmt.Errorf("cannot seal the encryption key: %v", err)
+
+	if err := sealRunObjectKeys(key, pbc, authKey, roleToBlName); err != nil {
+		return err
+	}
+
+	if err := sealFallbackObjectKeys(key, saveKey, rpbc, authKey, roleToBlName); err != nil {
+		return err
 	}
 
 	if err := stampSealedKeys(InstallHostWritableDir); err != nil {
 		return err
 	}
 
 	installBootChainsPath := bootChainsFileUnder(InstallHostWritableDir)
-	if err := writeBootChains(pbc, installBootChainsPath, 0); err != nil {
+	if err := writeBootChains(pbc, rpbc, installBootChainsPath, 0, 0); err != nil {
 		return err
 	}
 
 	return nil
 }
 
+func sealRunObjectKeys(key secboot.EncryptionKey, pbc predictableBootChains, authKey *ecdsa.PrivateKey, roleToBlName map[bootloader.Role]string) error {
+	modelParams, err := sealKeyModelParams(pbc, roleToBlName)
+	if err != nil {
+		return fmt.Errorf("cannot prepare for key sealing: %v", err)
+	}
+	sealKeyParams := &secboot.SealKeyParams{
+		ModelParams:            modelParams,
+		TPMPolicyAuthKey:       authKey,
+		TPMPolicyAuthKeyFile:   filepath.Join(InstallHostFDEDataDir, "tpm-policy-auth-key"),
+		TPMLockoutAuthFile:     filepath.Join(InstallHostFDEDataDir, "tpm-lockout-auth"),
+		TPMProvision:           true,
+		PCRPolicyCounterHandle: secboot.RunObjectPCRPolicyCounterHandle,
+	}
+	// The run object contains only the ubuntu-data key and the ubuntu-save key
+	// is stored inside the encrypted data partition, so that the normal run
+	// path only unseals one object (unsealing is expensive).
+	keys := []secboot.SealKeyRequest{
+		{
+			Key:     key,
+			KeyFile: filepath.Join(InitramfsEncryptionKeyDir, "ubuntu-data.sealed-key"),
+		},
+	}
+	if err := secbootSealKey(keys, sealKeyParams); err != nil {
+		return fmt.Errorf("cannot seal the encryption keys: %v", err)
+	}
+
+	return nil
+}
+
+func sealFallbackObjectKeys(key, saveKey secboot.EncryptionKey, pbc predictableBootChains, authKey *ecdsa.PrivateKey, roleToBlName map[bootloader.Role]string) error {
+	// also seal the keys to the recovery bootchains as a fallback
+	modelParams, err := sealKeyModelParams(pbc, roleToBlName)
+	if err != nil {
+		return fmt.Errorf("cannot prepare for fallback key sealing: %v", err)
+	}
+	sealKeyParams := &secboot.SealKeyParams{
+		ModelParams:            modelParams,
+		TPMPolicyAuthKey:       authKey,
+		TPMPolicyAuthKeyFile:   filepath.Join(InstallHostFDEDataDir, "tpm-policy-auth-key"),
+		TPMLockoutAuthFile:     filepath.Join(InstallHostFDEDataDir, "tpm-lockout-auth"),
```

Comment:
the fact that we have the files again seems wrong to me, see my comment in secboot

Reply:
ah, if TPMProvision is fail TPMLockoutAuthFile is not used afaict... see also my comment in secboot

Reply:
Changed to not pass the auth key filename in the second call (and in this case the file won't be written again).